### PR TITLE
Handle ObjectLiteral

### LINKED ISSUE #36: Handle ObjectLiteral

### PR DESCRIPTION
**Pull request by turlando**
Depends on #35

Fix #36

### CODE REVIEW

**Comment by turlando** [comment]
The code calls them ObjectLiteral, the documentation calls them [records](https://github.com/purescript/documentation/blob/master/language/Records.md). Should the test go in Test.Literal.ObjectLiteral or should we keep it separated since it involves more related stuff (accessor, updater, ...)?

**Comment by turlando** [comment]
String printing has to be fixed too. Should we do it in this PR or make another one?

**Comment by f-f** [comment]
@turlando 

> String printing has to be fixed too. Should we do it in this PR or make another one?

Should be done in another PR, and we should merge that first

> Should the test go in Test.Literal.ObjectLiteral or should we keep it separated since it involves more related stuff (accessor, updater, ...)?

No preference, but smaller tests are usually better

**Comment by turlando** [comment]
>>String printing has to be fixed too. Should we do it in this PR or make another one?

>Should be done in another PR, and we should merge that first

I created another issue just for that #38

**Comment by turlando** [comment]
Rebased off #39

File: src/Language/PureScript/Scheme/CodeGen/Scheme.hs [hunk 1]
```diff
@@ -15,19 +15,45 @@ app name args = List ((Symbol name):args)
...
+  where
+    go :: (Text, SExpr) -> SExpr
+    go (var, expr') = List [Symbol var, expr']
+
+-- (cond (test expr) ... (else expr))
+condWithElse :: [(SExpr, SExpr)] -> SExpr -> SExpr
+condWithElse clauses elseExpr = cond (clauses ++ [(Symbol "else", elseExpr)])
+
+-- (cond (test expr) ...)
+cond :: [(SExpr, SExpr)] -> SExpr
```
Comment (f-f) [inline src/Language/PureScript/Scheme/CodeGen/Scheme.hs:51]: Is there any value in having a `cond` without an `else` clause?

If our implementation is wrong and we don't match on any of the clauses then this will presumably return `nil`, which sure is going to be _hard_ to debug?

I'd prefer if we'd always emit an `else` clause, where we'd throw an "internal error" when we don't have an `else` clause at the application level. In this way if our implementation is wrong we'd have a pointer at where things blow up.

In other words, I think `cond` should be implemented with `condWithElse`, where we pass an `else` case that throws.

**Review by f-f — APPROVED** [review]
One small comment, but otherwise great stuff!

File: src/Language/PureScript/Scheme/CodeGen/Scheme.hs (hunk 1 above)
Comment (turlando) [reply src/Language/PureScript/Scheme/CodeGen/Scheme.hs:51]: I agree with your suggestion to always use `condWithElse` instead of `cond`.

As of now we're caling `cond` (without else clause) only at [`Transpiler.hs:129`](https://github.com/turlando/purescm/blob/1f7f27f70da735134de1804b8ed9802610e1a45b/src/Language/PureScript/Scheme/CodeGen/Transpiler.hs#L129) when handling case expressions.

In this case I think we're safe to have an exception raised if we hit the else branch.

I would still keep `cond` in `Scheme.hs` though since the cond (without else) is effectively possible in Scheme and provides a tiny lower abstraction for `condWithElse`. Maybe we can make it private?

**Comment by turlando** [comment]
Some things I'm noticing looking the code again:

- binderToTest can't handle string literals (I missed it in the previous PR)
- Maybe write `let1` for uses such as `Transpiler.hs:115`.
- Maybe factor out hashmap creation in a separate function?

File: src/Language/PureScript/Scheme/CodeGen/Scheme.hs (hunk 1 above)
Comment (f-f) [reply src/Language/PureScript/Scheme/CodeGen/Scheme.hs:51]: > Maybe we can make it private?

I don't see the point in keeping an unsafe function even if private. What I mean by "I think `cond` should be implemented with `condWithElse`" (instead of the other way around as it is now) is this:

```haskell
condWithElse :: [(SExpr, SExpr)] -> Maybe SExpr -> SExpr
condWithElse clauses maybeElseExpr
  = app "cond"
  $ map (\(test, expr) -> List [test, expr])
  $ clauses <> case maybeElseExpr of
      Nothing -> [Symbol "raise", String "Internal error while handling expression etc etc"]
      Just elseExpr -> [Symbol "else", elseExpr]

cond :: [(SExpr, SExpr)] -> SExpr
cond clauses = condWithElse clauses Nothing
```

In this way we could export both `cond` and `condWithElse`, and they would both be safe :slightly_smiling_face:

**Comment by f-f** [comment]
> 1. binderToTest can't handle string literals (I missed it in the previous PR)
> 2. Maybe write let1 for uses such as Transpiler.hs:115?
> 3. Maybe factor out hashmap creation in a separate function?

Of all of these I think only (1) is critical and should be fixed ASAP, the others I don't have an opinion on but you can go ahead and fix them here if you'd like

**Comment by turlando** [comment]
The first issue should be fixed with #40.

**Comment by f-f** [comment]
Great!

Then I think we are good to merge this after we take care of:
- rebasing/conflicts
- the `cond` stuff
